src/lsm/tree_fuzz.zig

```diff
@@ -541,21 +543,30 @@ fn EnvironmentType(comptime table_usage: TableUsage) type {
                     },
                     .put => |value| {
                         if (table_usage == .secondary_index) {
-                            if (model.get(Value.key_from_value(&value))) |old_value| {
-                                // Not allowed to put a present key without removing the old value first.
-                                env.tree.remove(&old_value);
+                            // Secondary index requires that the key implies the value (typically
+                            // key ≡ value), and that there are no updates.
+                            const canonical_value: Value = .{
+                                .id = value.id,
+                                .value = 0,
+                                .tombstone = value.tombstone,
+                            };
+                            if (model.contains(Value.key_from_value(&canonical_value))) {
+                                env.tree.remove(&canonical_value);
```

Comment:
Missed `// secondary_index only preserves keys - may return old values` subtelty during the initial impl, and decided that it's better to account for then when generating input data.

I am also wondering if we should add asserts to make the old code blow up at runtime?

Here: https://github.com/tigerbeetle/tigerbeetle/blob/fbda8466dcff2c64c7b7ee23d670183d7a9c72e3/src/lsm/compaction.zig#L1581-L1582

we could assert that tombstones are different, but everything _but_ tombstones is the same.

Reply:
:thinking: probably isn't worth it: in our real usage, Key=Value, so it's actually impossible to get divergance.